README.md
<h1 align="center">Mak<sup>mini<sup></h1>

<h5 align="center">A (small) piece of paper. Build on top of <a href="https://github.com/inns/mak">Mak</a>.</h5>

<div align="center">
  <a href="https://github.com/inns/mak-mini/releases">
    <img src='https://img.shields.io/github/downloads/inns/mak-mini/total.svg'/>
  </a>
  <a href="https://github.com/inns/mak-mini/releases">
    <img src='https://img.shields.io/github/package-json/v/inns/mak-mini.svg'/>
  </a>
  <a href="https://mak.ink">
    <img src='https://img.shields.io/website-up-down-green-red/https/mak.ink.svg?label=mak.ink'/>
  </a>
</div>


<img src='assets/mak-mini.png' width=128 height=128 align='right'/>

## The App
[Download latest releases](https://github.com/inns/mak-mini/releases)  
[Changelog](https://github.com/inns/mak-mini/blob/master/CHANGELOG.md)

### “Maximally Minimal”

The simplicity of plain text, the power of Markdown. 

Besides taking notes, you can **interact** with your document to organize it. Checkboxes and links are clickable, drop files to open, code editor shortcuts... Which all make it so intuitive to use in your way. To-do list, story outlines or linked fragments, it's up to you.

For links start with a hashtag (#), it's just a **file** in Mak. You can either use Markdown links (`[](#notes)`) or select "List" in the tray menu to jump to that file.

There're powerful **commands**, of course. By typing `go:`, you can quick jump to any files or create a new one. And by typing `set:`, you can do customizations or run advanced commands, like **print / generate PDF** and **export data**.

Most importantly, each advanced feature does not break the experience of any basic functionalities. If you don't need some feature, it doesn't exist.

For more details & usage, please read [Mak: a universal notepad.](https://github.com/inns/mak).  
If you have any suggestions or feedbacks, feel free to open a issue.

<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" width=880 />

<img src='screenshot/mak.png' width=539 align='right'/>

### macOS

Mak<sup>mini</sup> is a menubar App, just as simple as [Mak](https://mak.ink). But unlike the browser version, Mak<sup>mini</sup> is **NOT** designed to be focused on. The key point is **easy to access**.

You can use it just like stickers, to-do lists, quick notes, a temporary clipboard with read-it-later URLs, or **any mess you don't know where to put**.

This is not a focus writing tool. The Mak read/split preview mode are disabled currently.

<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" width=880 />

<img src='screenshot/touchbar.png' width=650 align='right'/>

### TouchBar Support (MacBook 2016)

TouchBar is convenient and intuitive, also a part of Mak's philosophy. Mak<sup>mini</sup> works perfectly with it.

<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" width=880 />

<img src='screenshot/windows-mak.png' width=531 align='right'/>

### Windows

There's also a Windows version. The dark theme looks gorgeous in Windows 10.

<img src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" width=880 />


## Share Mak<sup>mini</sup>
[![Twitter URL](https://img.shields.io/twitter/url/http/shields.io.svg?style=social)](https://twitter.com/intent/tweet?text=A+small+piece+of+paper:&url=https%3A%2F%2Fgithub.com%2Finns%2Fmak-mini)


## Development

Mak<sup>mini</sup> is built on top of Electron.

### Install dependencies
`$ yarn`

### Get started
Run `$ yarn dev`.  

When running in development mode, you can use the Inspector for debugging.

### Build Production
Run `$ yarn build` to generate a build for your environment (darwin/win32/etc.).  


## License & Acknowledgement
[TBD.]

By Shu (g@shud.in), SH 2017
https://shud.in


CHANGELOG.md
## 0.0.2 (Oct 9, 2017)
- Added env tags injection.
- Set App window size.

## 0.0.3 (Oct 10, 2017)
- Added icons.

## 0.0.4 (Oct 10, 2017)
- Added setting overwrites.

## 0.1.0 (Oct 18, 2017)
- Using mak.ink.
- New icon.
- Background color changes.
- Added contextmenu.
- Fixed keybindings.

## 0.1.1 (Oct 25, 2017)
- Added Mac TouchBar support.
- Fixed external URL issues.
- Re-arrange options.
- Added version info.
- Added readme and screenshots.

## 0.1.2 (Oct 25, 2017)
- Fixed tray icon in Windows.
- Fixed skipTaskbar in Windows.
- Code refactorings.

## 0.4.3 (May 31, 2018)
- Fixed broken Touchbar buttons.
- Fixed popup window & GitHub auth.
- Added Gist load to the tray menu.
- Removed `setInterval`.
- Refined Touchbar icons (current file, theme, line number, font size).
- New icon.
- Updated README.md.
- Updated version number (same as Mak).


src/index.js
const path = require('path')
const platform = require('os').platform()
const menubar = require('menubar')
const { BrowserWindow, Menu, shell } = require('electron')

const addTouchbar = require('./touchbar')
const addContextmenu = require('./contextmenu')
const { bindMak } = require('./status')

const { IS_DEV } = require('./env')
const { URL_MAK, URL_AUTH } = require('./config')
const WIDTH = 400
const HEIGHT = 600

const mb = menubar({
  icon: path.join(__dirname, '..', 'assets', `${platform}-IconTemplate.png`),
  index: URL_MAK,
  preloadWindow: true,
  width: WIDTH,
  height: HEIGHT,
  backgroundColor: 'white',
  // alwaysOnTop: true
})

mb.on('ready', function() {
  mb.window.setSkipTaskbar(true)

  // Workaround to fix window position when statusbar at top for win32
  if (platform === 'win32') {
    if (mb.tray.getBounds().y < 5) {
      mb.setOption('windowPosition', 'trayCenter')
    }
  }

  mb.window.webContents.on('did-finish-load', function() {
    mb.window.webContents.executeJavaScript(';window.IS_MAK_MINI = true;')
    // setInterval(tick, 500)
  })

  mb.window.webContents.on('will-navigate', function(event, url) {
    if (!url.startsWith(URL_MAK)) {
      // external links
      event.preventDefault()
      shell.openExternal(url)
    }
  })

  mb.window.webContents.on('new-window', function(event, url, frameName, disposition, options, additionalFeatures) {
    if (url.startsWith(URL_AUTH) || url.startsWith(URL_MAK)) {
      event.preventDefault()
      Object.assign(options, {
        modal: false,
        frame: true,
        parent: mb.window,
      })

      let window = new BrowserWindow(options)
      event.newGuest = window

      mb.setOption('alwaysOnTop', true)

      window.loadURL(url)
      window.show()

      // Trick to keep the menubar app focused
      window.on('close', () => mb.setOption('alwaysOnTop', false))

      return
    }

    event.preventDefault()
    shell.openExternal(url)
  })

  bindMak(mb)
  addContextmenu(mb)
  addTouchbar(mb)

  // mb.app.commandLine.appendSwitch('--enable-experimental-web-platform-features')
})


src/status.js
const status = {}
const listeners = {}

function bindMak(menubar) {
  const execJS = (...args) => menubar.window.webContents.executeJavaScript(...args)

  const registerEditorEv = (name) => {
    // Trick:
    // subscribe the stream only once since we can not use callback 
    // but only Promise here in Electron.
    execJS(`new Promise(res => 
      window.Mak.EditorSubject
        .filter(ev => ev.type === '${name}')
        .first()
        .subscribe(res)
    )`)
    .then((res) => {
      // Callbacks
      status[name] = res.data
      listeners[name] && listeners[name].forEach(fn => fn(status))
      // Register again
      registerEditorEv(name)
    })
  }
  
  // Init values after Mak instance loaded
  execJS('window.Mak.mak && window.Mak.mak.getSettings()', settings => {
    status.initSettings = settings
    listeners.initSettings && listeners.initSettings.forEach(fn => fn(status))
  })

  // Bind events
  registerEditorEv('afterLoadSettings')
  registerEditorEv('title')
}

function on(type, fn) {
  listeners[type] = [...(listeners[type] || []), fn]
}

module.exports = {
  status,
  on,
  bindMak
}


src/touchbar.js
const path = require('path')
const { TouchBar } = require('electron')

const emit = require('./cmd')
const { on, status } = require('./status')

const { URL_MAK } = require('./config')

const {
  TouchBarLabel,
  TouchBarButton,
  TouchBarSpacer,
  TouchBarGroup
} = TouchBar

module.exports = function addTouchbar (menubar) {
  const execJS = (...args) => menubar.window.webContents.executeJavaScript(...args)
  const DARK_ICON = path.join(__dirname, '..', 'assets', `moon@4x.png`)
  const LIGHT_ICON = path.join(__dirname, '..', 'assets', `sun@4x.png`)
  const FS_ICON = path.join(__dirname, '..', 'assets', `font@4x.png`)
  const NU_ICON = path.join(__dirname, '..', 'assets', `list-numbers@4x.png`)

  let currentSettings = {}

  const homeBtn = new TouchBarButton({
    label: '#',
    click: () => execJS(`;window.location = '${URL_MAK}#';`)
  })
  
  const listBtn = new TouchBarButton({
    label: '#list',
    click: () => execJS(`;window.location = '${URL_MAK}#list';`)
  })
  
  const preferencesBtn = new TouchBarButton({
    label: '#preferences',
    click: () => execJS(`;window.location = '${URL_MAK}#preferences';`)
  })
  
  const fontsizeBtn = new TouchBarButton({
    icon: FS_ICON,
    backgroundColor: '#000',
    click: () =>
      emit(menubar, 'CMDSubject', {
        name: 'fs',
        arg: currentSettings.fs === 'normal' ? 'large' : (currentSettings.fs === 'large' ? 'small' : 'normal')
      })
  })

  const linenumberBtn = new TouchBarButton({
    icon: NU_ICON,
    backgroundColor: '#000',
    click: () =>
      emit(menubar, 'CMDSubject', {
        name: currentSettings.nu ? 'nonum' : 'num'
      })
  })

  const themeBtn = new TouchBarButton({
    icon: DARK_ICON,
    backgroundColor: '#000',
    click: () =>
      emit(menubar, 'CMDSubject', {
        name: 'theme',
        arg: currentSettings.theme === 'light' ? 'dark' : 'light'
      })
  })
  
  const makBtn = new TouchBarButton({
    label: 'Mak',
    backgroundColor: '#000'
  })
  
  const titleBtn = new TouchBarButton({
    label: '#',
    backgroundColor: '#000'
  })
  
  const touchBar = new TouchBar({
    items: [
      titleBtn,
      new TouchBarSpacer({ size: 'flexible' }),
      themeBtn,
      linenumberBtn,
      fontsizeBtn,
      new TouchBarGroup({
        items: [homeBtn, listBtn, preferencesBtn]
      })
    ],
    escapeItem: makBtn
  })

  const updateTouchbar = (type, data) => {
    if (type === 'settings') {
      themeBtn.icon = data.theme === 'light' ? DARK_ICON : LIGHT_ICON
      currentSettings = data
    } else if (type === 'title') {
      let originalTitle = '#' + data.toLowerCase()
      titleBtn.label = originalTitle.length <= 15 ? originalTitle : (originalTitle.slice(0, 15) + '...')
    }
  }

  on('initSettings', status => {
    updateTouchbar('settings', status.initSettings)
  })
  on('afterLoadSettings', status => {
    updateTouchbar('settings', status.afterLoadSettings)
  })
  on('title', status => {
    updateTouchbar('title', status.title)
  })
  menubar.window.setTouchBar(touchBar)  
}


src/contextmenu.js
const platform = require('os').platform()
const { Menu, shell, clipboard, dialog } = require('electron')

const { IS_DEV } = require('./env')
const VERSION = require('../package.json').version
const { URL_MAK, URL_MAK_GITHUB, URL_GIST } = require('./config')

module.exports = function addContextmenu (menubar) {
  const openExternal = URL => shell.openExternal(URL)
  const execJS = (...args) => menubar.window.webContents.executeJavaScript(...args)

  const shortCuts = [
    {
      label: 'Mak',
      submenu: [
        { label: 'Quit', accelerator: 'Command+Q', click: () => menubar.app.quit() },
        { label: 'Undo', accelerator: 'CmdOrCtrl+Z', selector: 'undo:' },
        { label: 'Redo', accelerator: 'Shift+CmdOrCtrl+Z', selector: 'redo:' },
        { label: 'Cut', accelerator: 'CmdOrCtrl+X', selector: 'cut:' },
        { label: 'Copy', accelerator: 'CmdOrCtrl+C', selector: 'copy:' },
        { label: 'Paste', accelerator: 'CmdOrCtrl+V', selector: 'paste:' },
        {
          label: 'Select All',
          accelerator: 'CmdOrCtrl+A',
          selector: 'selectAll:'
        }
      ]
    }
  ]
  
  let gistId = null
  let template = [
  {
    label: 'Open Mak in Browser',
    click: () => openExternal(URL_MAK)
  },
  { type: 'separator' },
  {
    label: 'Home',
    click: () => {
      execJS(`;window.location = '${URL_MAK}';`)
      menubar.showWindow()
    }
  },
  {
    label: 'List',
    click: () => {
      execJS(`;window.location = '${URL_MAK}#list';`)
      menubar.showWindow()
    }
  },
  {
    label: 'Preferences',
    click: () => {
      execJS(`;window.location = '${URL_MAK}#preferences';`)
      menubar.showWindow()
    }
  },
  { type: 'separator' },
  {
    label: `About Mak v${VERSION}`,
    click: () => openExternal(URL_MAK_GITHUB)
  },
  { type: 'separator' },
  {
    label: `Open a Gist...`,
    click: () => {
      if (!gistId) {
        dialog.showMessageBox({
          type: 'info',
          message: 'Please copy a Gist URL to open.',
          buttons: ['OK', 'Browse Gists...'],
          cancelId: 0,
        }, res => {
          if (res === 1) {
            openExternal(URL_GIST)
          }
        })
      } else {
        execJS(`;window.location = '${URL_MAK}#?gist=${gistId}';`)
        menubar.showWindow()
      }
    }
  },
  ...(IS_DEV
    ? [
        { type: 'separator' },
        {
          label: '*Dev ' + platform
        },
        {
          label: '*Inspect',
          click: () => menubar.window.openDevTools()
        },
        { type: 'separator' }
      ]
    : [{ type: 'separator' }]),
    { label: 'Quit', role: 'quit' }
  ]

  menubar.tray.on('right-click', function() {
    let str = (clipboard.readText() || '').trim()
    if (str.startsWith(URL_GIST) || /^[a-z0-9]+$/.test(str)) {
      // a gist
      gistId = str.split('/').filter(s => s).pop()

      let displayGistId = gistId.length <= 15 ? gistId : (gistId.slice(0, 15) + '...')
      template[8].label = `Open ${displayGistId}`
    } else {
      gistId = null
      template[8].label = 'Open a Gist...'
    }
    const contextMenu = Menu.buildFromTemplate(template)
  
    Menu.setApplicationMenu(contextMenu)
    menubar.tray.popUpContextMenu(contextMenu)
  })
  Menu.setApplicationMenu(Menu.buildFromTemplate(shortCuts))
}


main.js
const path = require('path')
const platform = require('os').platform()

const menubar = require('menubar')

const env = process.env.NODE_ENV
const isDev = env && env.trim() === 'development'

const { Menu, shell, TouchBar } = require('electron')
const {
  TouchBarLabel,
  TouchBarButton,
  TouchBarSpacer,
  TouchBarGroup
} = TouchBar

const INDEX = 'https://mak.ink'
const WIDTH = 400
const HEIGHT = 600
const VERSION = require('./package.json').version

const openExternal = URI => shell.openExternal(URI)
const execJS = (...args) =>
  mb.window && mb.window.webContents.executeJavaScript(...args)

var mb = menubar({
  icon: path.join(__dirname, 'assets', `${platform}-IconTemplate.png`),
  index: INDEX,
  preloadWindow: true,
  width: WIDTH,
  height: HEIGHT,
  backgroundColor: 'white'
})

const contextMenu = Menu.buildFromTemplate([
  {
    label: 'Mak in Browser',
    click: () => openExternal('https://mak.ink')
  },
  { type: 'separator' },
  {
    label: 'Home',
    click: () => {
      execJS(`;window.location = '${INDEX}';`)
      mb.showWindow()
    }
  },
  {
    label: 'List',
    click: () => {
      execJS(`;window.location = '${INDEX}#list';`)
      mb.showWindow()
    }
  },
  {
    label: 'Preferences',
    click: () => {
      execJS(`;window.location = '${INDEX}#preferences';`)
      mb.showWindow()
    }
  },
  { type: 'separator' },
  {
    label: `About Mak v${VERSION}`,
    click: () => openExternal('https://github.com/inns/mak')
  },
  ...(isDev
    ? [
        { type: 'separator' },
        {
          label: 'Development ' + platform
        },
        {
          label: 'Inspect',
          click: () => mb.window.openDevTools()
        },
        { type: 'separator' }
      ]
    : [{ type: 'separator' }]),
  { label: 'Quit', role: 'quit' }
])

const homeBtn = new TouchBarButton({
  label: '#',
  click: () => execJS(`;window.location = '${INDEX}#';`)
})

const listBtn = new TouchBarButton({
  label: '#list',
  click: () => execJS(`;window.location = '${INDEX}#list';`)
})

const preferencesBtn = new TouchBarButton({
  label: '#preferences',
  click: () => execJS(`;window.location = '${INDEX}#preferences';`)
})

const themeBtn = new TouchBarButton({
  label: '🌖',
  backgroundColor: '#000',
  click: () =>
    execJS(`
    ;window[window.settings.theme === 'light' ? 'dark' : 'light']();
  `)
})

const cntBtn = new TouchBarButton({
  label: '✍️',
  backgroundColor: '#000',
  click: () => execJS(`;window[window.settings.cnt ? 'nocnt' : 'cnt']();`)
})

const titleBtn = new TouchBarButton({
  label: 'Mak',
  backgroundColor: '#000',
  click: () => execJS(`;window[window.settings.cnt ? 'nocnt' : 'cnt']();`)
})

const touchBar = new TouchBar({
  items: [
    new TouchBarSpacer({ size: 'small' }),
    new TouchBarGroup({
      items: [homeBtn, listBtn, preferencesBtn]
    }),
    new TouchBarSpacer({ size: 'small' }),
    themeBtn,
    cntBtn,
    new TouchBarSpacer({ size: 'small' })
  ],
  escapeItem: titleBtn
})

const tick = () => {
  if (mb.window) {
    mb.window.webContents.executeJavaScript(
      'window.document.title',
      false,
      title => {
        titleBtn.label = title
      }
    )
  }
}

const shortCuts = [
  {
    label: 'Mak',
    submenu: [
      { label: 'Quit', accelerator: 'Command+Q', click: () => mb.app.quit() },
      { label: 'Undo', accelerator: 'CmdOrCtrl+Z', selector: 'undo:' },
      { label: 'Redo', accelerator: 'Shift+CmdOrCtrl+Z', selector: 'redo:' },
      { label: 'Cut', accelerator: 'CmdOrCtrl+X', selector: 'cut:' },
      { label: 'Copy', accelerator: 'CmdOrCtrl+C', selector: 'copy:' },
      { label: 'Paste', accelerator: 'CmdOrCtrl+V', selector: 'paste:' },
      {
        label: 'Select All',
        accelerator: 'CmdOrCtrl+A',
        selector: 'selectAll:'
      }
    ]
  }
]

mb.on('ready', function() {
  mb.window.setSkipTaskbar(true)

  mb.window.webContents.on('did-finish-load', function() {
    mb.window.webContents.executeJavaScript(';window.IS_MAK_MINI = true;')
    setInterval(tick, 500)
  })

  mb.window.webContents.on('will-navigate', function(event, url) {
    if (!url.startsWith(INDEX)) {
      // external links
      event.preventDefault()
      shell.openExternal(url)
    }
  })

  mb.tray.on('right-click', function() {
    mb.tray.popUpContextMenu(contextMenu)
  })

  mb.window.setTouchBar(touchBar)

  Menu.setApplicationMenu(Menu.buildFromTemplate(shortCuts))
  // mb.app.commandLine.appendSwitch('--enable-experimental-web-platform-features')
})
